packages/utreexo/src/stump/proof.cairo

```diff
@@ -27,3 +30,173 @@ impl UtreexoBatchProofDisplay of Display<UtreexoBatchProof> {
+                row_len_acc += row_len;
+                row_len /= 2;
+                actual_row_len /= 2;
+
+                if row_len == 0 {
+                    inner_result =
+                        Result::Err(
+                            format!("Position {pos} is out of the forest range {row_len_acc}.")
+                        );
+                    break;
+                }
+            };
+
+            // If row length is odd and we are at the edge this is a root.
+            if pos == row_len_acc + actual_row_len - 1 && actual_row_len % 2 == 1 {
+                calculated_root_hashes.append(node);
+                row_len_acc += row_len;
+                row_len /= 2;
+                actual_row_len /= 2;
+                continue;
+            };
+
+            let parent_node = if (pos - row_len_acc) % 2 == 0 {
+                // Right sibling can be both leaf/computed or proof.
+                let right_sibling = if next_leaf_pos == pos + 1 {
+                    let res = next_leaf;
+                    let (a, b) = leaf_nodes.pop_front().unwrap_or((Bounded::<u64>::MAX, 0));
+                    next_leaf_pos = a;
+                    next_leaf = b;
+                    res
+                } else if next_computed_pos == pos + 1 {
+                    let res = next_computed;
+                    let (a, b) = computed_nodes.pop_front().unwrap_or((Bounded::<u64>::MAX, 0));
+                    next_computed_pos = a;
+                    next_computed = b;
+                    res
+                } else {
+                    if sibling_nodes.is_empty() {
+                        inner_result = Result::Err("Proof is empty.");
+                        break;
+                    };
+                    sibling_nodes.pop_front().unwrap()
+                };
+                parent_hash(node, right_sibling)
+            } else {
+                // Left sibling always from proof.
+                if let Option::Some(left_sibling) = sibling_nodes.pop_front() {
+                    parent_hash(left_sibling, node)
+                } else {
+                    inner_result = Result::Err("Proof is empty.");
+                    break;
+                }
+            };
+
+            let parent_pos = row_len_acc + row_len + (pos - row_len_acc) / 2;
+
+            if next_computed_pos == Bounded::<u64>::MAX {
+                next_computed_pos = parent_pos;
+                next_computed = parent_node;
+            } else {
+                computed_nodes.append((parent_pos, parent_node));
+            }
+        };
+
+        if !sibling_nodes.is_empty() {
+            return Result::Err("Proof should be empty");
+        }
+
+        if inner_result != Result::Ok((array![].span())) {
+            inner_result
+        } else {
+            Result::Ok((calculated_root_hashes.span()))
+        }
+    }
+}
+
+/// Computes the next power of two of a u64 variable.
+fn next_power_of_two(mut n: u64) -> u64 {
```

Comment:
Let's put this to utils and add couple of unit tests (just to be sure)?

Reply:
sure